source/SIL.AppBuilder.Portal.Frontend/src/lib/string/utils.ts

```diff
@@ -0,0 +1,51 @@
+const TOKEN_REGEX = /(|\s)(\w+):/;
+
+export interface IParsedTokens {
+  [key: string]: any;
+}
+
+// example:
+// 1.
+//   'name:"hi" language:English*'
+//   => ['name', 'hi', ' ', 'language', 'English*']
+//
+// 2.
+//   'name:"hi" language:'
+//   => ['name', 'hi', ' ', 'language']
+//
+// 3.
+//   'name:"hi" language: something:'
+//   => ['name', 'hi', ' ', 'language', ' ', 'something']
+export function tokensToObject(search: string): IParsedTokens {
+  if (search === '') {
+    return {};
+  }
+  const tokens: string[] = search
+    .split(TOKEN_REGEX)
+    .filter(Boolean)
+    .map(t => t.replace(/(^"|"$)/g, ""));
+
+  const result: IParsedTokens = {};
+
+  for(let i = 0; i < tokens.length; i++) {
```

Comment:
I was going to say that you could do this using a reducer (more elegant) but then I notice you need to control the index. Now that think about it... you can still skip odd indexes.

Reply:
I can't skip odd indexes, because sometimes there isn't always a value. at the bottom of the for loop I do skip over an index IFF that next index was a value. :)